drivers/ve-direct.c

```diff
@@ -0,0 +1,298 @@
+		if (ret < 0)
+			return STAT_INSTCMD_FAILED;
+		r_start += ret;
+		// check if we read full line with the reply..
+		reply = strchr(line, ':');
+		endl = NULL;
+		if (reply != NULL)
+			endl = strchr(reply, '\n');
+		// already seen start of command reply
+		if (reply != NULL && endl != NULL)
+			break;
+
+		process_text_buffer();
+	}
+
+	// process what's left
+	while (line[0] != ':')
+	{
+		if (process_text_buffer() < 0)
+			break;
+	}
+
+	endl = NULL;
+	endl = strchr(line, '\n');
+
+	if (endl != NULL)
+	{
+		*endl = '\0';
+		upsdebugx(2, "reply to command: %s", line);
+		int checksum = -1;
+		sscanf(endl - 2, "%02X", &checksum);
+		endl[-2] = '\0';
+
+
+		if (checksum == ve_checksum(line[1], line + 2))
+		{
+			upsdebugx(3, "correct checksum on reply to command %c", line[1]);
+		}
+		else
+		{
+			upslogx(1, "invalid checksum on reply to command %c", line[1]);
+			return STAT_INSTCMD_FAILED;
+		}
+
+
+		if (ve_return != NULL)
+		{
+			memset(ve_return, 9, ret_lenght);
+
+			size_t r_len = endl - line - 4;
+			if (r_len > 0)
+			{
+				strncpy(ve_return, line + 2, r_len > ret_lenght ? ret_lenght : r_len);
+			}
+		}
+
+		ve_copy (endl + 1);
+	}
+
+	return STAT_INSTCMD_HANDLED;
+}
+
+static int instcmd(const char *cmdname, const char *extra)
+{
+	if (!strcasecmp(cmdname, "ve-direct.ping"))
+		return ve_command('1',NULL,NULL,0);
+	if (!strcasecmp(cmdname, "ve-direct.get"))
+		return ve_command(VE_GET,extra,NULL,0);
+	if (!strcasecmp(cmdname, "ve-direct.set"))
+		return ve_command(VE_SET,extra,NULL,0);
+	if (!strcasecmp(cmdname, "beeper.disable"))
+		return ve_command(VE_SET,"FCEE0000",NULL,0);
+	if (!strcasecmp(cmdname, "beeper.enable"))
+		return ve_command(VE_SET, "FCEE0001",NULL,0);
+	upsdebugx(1, "instcmd: unknown command: %s", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+	
+void upsdrv_initinfo(void)
+{
+	dstate_addcmd("ve-direct.ping");
+	dstate_addcmd("ve-direct.set");
+	dstate_addcmd("ve-direct.get");
+
+	dstate_addcmd("beeper.disable");
+	dstate_addcmd("beeper.enable");
+
+	upsh.instcmd = instcmd;
+}
+
+void upsdrv_updateinfo(void)
+{
+	/* ups.status */ 
+	status_init();
+	status_set("OL");
+	status_commit();
+
+	upsdebugx(1, "upsdrv_updateinfo: ups.status = %s", dstate_getinfo("ups.status"));
+
+	/************** ups.x ***************************/
+	while (1)
+	{
+		if (r_start >= sizeof(line))
+			r_start = 0;
+		memset(line + r_start, 0, sizeof(line) - r_start);
+		int nred = ser_get_buf(upsfd, line + r_start, sizeof(line) - r_start, 0, 200);
+		if (nred < 0)
+			break;
+		if (nred == 0)
+			continue;
+		r_start += nred;
+
+		while (process_text_buffer() >= 0)
+		{
+		}
+
+		break;
+	}
+
+	dstate_dataok();
+}
+
+void upsdrv_shutdown(void)
+{
+}
+
+void upsdrv_help(void)
+{
+	printf("Read The Fine Manual ('man 8 ve-direct')\n");
+}
+
+/* list flags and values that you want to receive via -x */
+void upsdrv_makevartable(void)
+{
+}
+
+void upsdrv_initups(void)
+{
+#ifndef TESTING
```

Comment:
The `TESTING` #define is not handled elsewhere in NUT, but if you include it, I would recommend adding a case to parse a few static text buffers. This will help other developers who might have to maintain the code later down the road.